### ISSUE TITLE
What are Web Assembly threads?

### ISSUE THREAD

**Issue by jfbastien** [WebAssembly/design#104]
Are Web Assembly specified in terms of [WebWorkers](https://developer.mozilla.org/en-US/docs/Web/API/Web_Workers_API/Using_web_workers), or are they different? What are the differences, or are WebWorkers just an implementation detail?

Maybe the polyfill is OK with WebWorkers, and Web Assembly does its own thing that's closer to pthreads.

We need to ensure that Web Assembly can work outside of a browser.

**Comment by trevnorris** [comment]
If I may offer a few thoughts from the server-side of things (specifically node). There are three main use cases for threads (there are probably "proper" names for these, but you'll get the idea):

- Computationally heavy native work. These have no JS stack. Values are passed, a callback is set and when the numbers have been crunched the callback receives the return value.
- Immediately exiting thread. A single file is passed, along with optional arguments, which runs to completion, during which time return values can be messaged back to the parent thread. This one has two variations.
  - No I/O access thread. The spawned thread is essentially synchronous, but to further clamp down on the scope of work these threads can do they are not allowed to do any sync I/O operations either.
  - I/O thread. Allowed to require other modules and perform sync I/O.
- Event loop thread. These are essentially just like a running process. Except in the way they communicate, and have the ability to share memory.

These have variations on how developers want them implemented. Like does the immediately exiting thread create a new JS stack every time, or use a new one. Is the thread closed when the operation is complete or reused. And also is a thread, sync or async, joined causing the main thread to hault until the spawned thread is complete.

Sorry if this was too many examples. What I'm getting at is it seems like wasm's ability to work with threads will either not be low level enough or extensive enough to fit these needs (I don't expect it could concerning the first example). Leaving server applications needing to use their own bindings. This correct?

**Comment by sunfishcode** [comment]
I don't have all the answers, but one thing I can comment on is that WebAssembly threads are going to have access to a pthreads-level API, so applications will have quite a lot of control. Decisions like when and how to use pthread_join are largely determined by the application (or by a library linked into the application).

Also, WebAssembly itself is being designed to be independent of JS, and it will be possible to have WebAssembly threads with no JS code on the stack.

**Comment by trevnorris** [comment]
Thank you. That's excellent to hear.

**Comment by sunfishcode** [comment]
What things here do we need to decide on for the MVP, and what things can wait until we actually introduce shared memory?

**Comment by lukewagner** [comment]
I don't see it directly blocking anything in the MVP.  That being said, for something so central, it seems like we should have a pretty clear idea about the feature (and some experimental experience) before the MVP is actually finalized/released.  I don't see "blocks_binary", though.

**Comment by jfbastien** [comment]
A few issues on WebWorkers detailed here: https://github.com/lars-t-hansen/ecmascript_sharedmem/issues/2

**Comment by lukewagner** [comment]
FWIW, both of those are impl issues and ones that we're expecting to address in FF.

**Comment by lars-t-hansen** [comment]
@lukewagner, they are not entirely implementation issues.  The worker startup semantics are allowed by the worker spec.  The limitation on the number of threads is a consequence of wanting to prevent DOS attacks, but it's very crude and something better, with observability (ie exception thrown on failure to create a worker), would be welcome by many applications, but probably requires a spec change too.  Additionally, so far as I can tell there's no way to detect if a worker has gone away, which is a bizarre spec hole, but I've not found a mechanism for it.

**Comment by lukewagner** [comment]
@lars-t-hansen As for workers not starting before returning to the event loop, when you say "allowed by the spec", do you just mean by the spec not specifying when progress is made or does it specifically mention this case?  As for the limitation on number of threads, you're right, what is needed (in addition to a higher quota) is some visible error to indicate the quota is exceeded.

**Comment by jfbastien** [comment]
I think a forward-progress guarantee is what we want here, in line with [Torvald Riegel's N4439 paper to the C++ standards committee](http://wg21.link/n4439).

**Comment by lars-t-hansen** [comment]
@lukewagner, my memory is that the spec has no language addressing progress at all.  Initially I thought it didn't matter (much) since it was not observable to the creating agent without that agent waiting for an event, but that's not really so - if the worker has access to XHR or rendering then the main thread could fork off a worker to have an external effect, and the absence of the effect would be visible to a remote web server or to the user.

@jfbastien, I have not read that paper yet but I agree, something like that is desirable.

(I'll write to our Web standards people re the Worker spec.)

**Comment by lars-t-hansen** [comment]
A couple more notes on workers.  The central section is WHATWG spec 10.2.4, Processing Model

* The "run a worker" algorithm: "1. Create a separate parallel execution environment (i.e. a separate thread or process or equivalent construct), and run the rest of these steps in that context."). This implies fairly strongly ("parallel", "thread or process") that a worker requires true concurrency, prohibiting both the return-to-event-loop delay and the indefinite wait-for-resource delay.
* The "kill a worker" algorithm has an interesting note: "User agents may invoke the "kill a worker" processing model on a worker at any time", ie, a worker can just die for no reason at all.  (And it's not obvious that this is directly observable.)

**Comment by lukewagner** [comment]
@lars-t-hansen Hah, good point regarding observability via network (particularly when you take into account sync XHR on the main thread) >:)

Regarding the "at any time" "kill a worker" part, that's weird.  I expect it's referring to the slow-script dialog, but, if so, it seems useful to better define that as part of a larger application event that kills all the workers and scripts.

**Comment by lars-t-hansen** [comment]
@lukewagner, The service worker spec (https://slightlyoff.github.io/ServiceWorker/spec/service_worker/) provides minimal justification for the "kill a worker" behavior (indeed, lack of UI for a slow script dialog, see the section "Lifetime") but no actual guidance on how to avoid being gunned down.  For a computational worker in a SAB context that license to kill is particularly troublesome, as the default mode for such a worker will be that it waits on a condition variable for more work, not that it returns to its event loop.

**Comment by lars-t-hansen** [comment]
Bug filed against the WHATWG spec here: https://www.w3.org/Bugs/Public/show_bug.cgi?id=29039.

**Comment by jfbastien** [comment]
@slightlyoff can probably chime in on web worker + service worker and the "license to kill".

**Comment by emanuelpalm** [comment]
I guess running webasm in strictly single-process environments would be doable even if the spec would require the presence of a pthreads-level API? As parallelism, as opposed to concurrency, cannot be guaranteed, an implementation would be free to simply treat created "pthreads" as blocks to schedule arbitrarily?

I guess no stricter guarantees than soft realtime will be promised?

**Comment by jfbastien** [comment]
@emanuelpalm agreed, a valid implementation could emulate a single-processor system.

Hard-realtime isn't something I think we can guarantee in general because WebAssembly doesn't know a-priori which machine it'll execute on. I see this limitation as similar to promising constant-time algorithms: the `.wasm` file doesn't know how it'll actually get lowered and exactly what guarantees it can expect from the JIT and machine it'll execute on.

**Comment by jbondc** [comment]
I'm really excited about this:
http://images.nvidia.com/events/sc15/SC5105-open-source-cuda-compiler.html
https://developer.nvidia.com/CUDA-LLVM-Compiler

Any thoughts if implementing a ```forall``` parallel loop would be feasible in Web Assembly?

What could the pipeline look like as a concrete way to experiment?
```web assembly -> wasm binary -> chrome -> gpu```

Seems more convenient to have something like:
```web assembly -> binary (jit?) -> chrome -> gpu```

**Comment by jfbastien** [comment]
@jbondc I think your thinking is closer to the C++ parallelism TS, which requires runtime support. It would be possible, but WebAssembly currently doesn't have ongoing work for GPUs. It's important, but not the primary focus for MVP. The work by Robert's team is pretty different from what WebAssembly would do (though it could benefit from that work).

**Comment by jbondc** [comment]
Yes this looks right:
https://github.com/cplusplus/parallelism-ts/blob/master/parallelism-ts.pdf

But more interested in writing my own language that compiles down to Web Assembly (and has parallelism support).

**Comment by lukewagner** [comment]
I think the best thing wasm can and should do is to provide the raw hardware primitives for parallelism (threads and SIMD) within the [general-purpose CPU model assumed by WebAssembly](https://github.com/WebAssembly/design/blob/master/Portability.md#assumptions-for-efficient-execution) to the language/library/tool authors, leaving existing and future Web APIs to access hardware outside the general-purpose CPU (like WebGL for the GPU).  Then the language/library/tool authors can build up abstractions that target a specific model of parallelism.  (This is basically just an [Extensible Web ](https://extensiblewebmanifesto.org/) argument.)  Based on this, I would think the libraries in the C++ parallelism TS would be provided as libraries on top of the abovementioned primitives of the web platform, not as builtin functions (at least, not in the near-term).

**Comment by jbondc** [comment]
Thoughts on something like this:
http://www.actor-framework.org/pdf/manual.pdf
https://github.com/actor-framework/actor-framework
https://www.youtube.com/watch?v=VOFOj3LLXRk

Is this a possible story?

**Comment by JSStats** [comment]
@jbondc It's a higher level abstraction and a framework that I expected could be implemented in wasm with good performance. If someone tried porting this and found some show stoppers then this might suggest extra low level threading support to be considered for wasm?

My understand behind the reasons for avoiding such high level abstractions is that they generally have niche uses and also they have inherent assumptions and limitations that do not map well to the hardware, so there might not be the general interest to warrant direct support but perhaps someone can rally support for some other parallel computation models.

**Comment by jbondc** [comment]
It's not a higher level abstaction. The "Actor Model" (https://en.wikipedia.org/wiki/Actor_model) is a different theoretical model to think about computation (e.g. vs. a finite turing machine). 

Based on hardware (e.g. gpu by Nvidia), you can implement in a VM some messaging passing actorLike thing. The way I see it, if either Chrome, Chakra, Firefox or Webkit would implement pattern matching + an actorLike thing, then we'd get C++ parallelism for free, memory sharing + threads, and pretty much any other concurrent model.

**Comment by jbondc** [comment]
Related in case someone wants to hack something togheter:
https://github.com/solodon4/Mach7

**Comment by JSStats** [comment]
@jbondc It's not SMP and it doubt it could implement SMP code with any level of efficiency? Consider how it would implement atomically updated stacks and queues and caches etc in a code written for shared memory? The planned wasm support should allow some of these 'Actor Model' programming languages to be implemented in wasm, and the atomic operations might be used to implement fast message queues and memory management to keep the actors apart etc.

**Comment by jbondc** [comment]
@JSStats I'm more trying to look at good building blocks. Better hardware is already here.

There's a good discussion about hardware here:
http://amturing.acm.org/acm_tcc_webcasts.cfm (Computer Architecture)

The threads / shared memory design is bad. 
Hewitt's ```Agent Model``` or some variation of it is the way to go.

As a concrete example, this is good work:
http://2014.rtss.org/wp-content/uploads/2014/12/RTSS2014-sifakis-distr.pdf

Page 34. Threads bad.
Page 48. Good model.

Page 59-60. A ~'binary jit' could apply here: Distributed BIP generator.

I'll try some experiements in the coming months, hopefully others will join.

**Comment by sunfishcode** [comment]
Moving this milestone to Post-MVP, so that the MVP milestone focuses on actionable tasks for the MVP. It's obviously important to plan for threads, but it's my impression that there's sufficient awareness already, so we don't need this issue to remind us. If people disagree, I'm fine moving it back.

@jbondc If you are planning to do some experiments, please open a new issue to report your findings. Thanks!

**Comment by jbondc** [comment]
@sunfishcode Will do, but largely gave up.  

For anyone interested in BIP: 
http://www-verimag.imag.fr/New-BIP-tools.html

Is the thinking then with WASM threads, that code like this would compile into atomic load/store ops?
```
class FastBlockableMutex {
 protected:
  atomic<bool> mIsLocked;
  atomic<int> mWaiters;
  atomic<int> mBlockers;
  mutex mMutex;
 
};
```

**Comment by jfbastien** [comment]
@jbondc it would need instructions to lower into ops ;-)
Yes, atomic accesses would as long as the size is lock-free. Mutex would likely go to an atomic and then some futex-like thing. I expect the model to be close to [this](https://github.com/tc39/ecmascript_sharedmem).

**Comment by malthe** [comment]
@jbondc isn't the idea to provide threads and compare-and-swap etc so that one can build a lock-free actor implementation on top of it?

**Comment by jbondc** [comment]
@malthe Ya, after some research that's what ponylang does:
https://github.com/ponylang/ponyc/blob/master/src/libponyrt/actor/actor.c
https://github.com/ponylang/ponyc/blob/master/src/common/pony/detail/atomics.h

There's also some interesting patterns of parallelism with HPX:
https://www.youtube.com/watch?v=6Z3_qaFYF84&feature=youtu.be&t=16m1s
https://github.com/STEllAR-GROUP/hpx/tree/master/external/atomic/boost/atomic/detail

And then GPU programming:
https://www.youtube.com/watch?v=KHa-OSrZPGo
https://github.com/ponylang/ponyc/issues/1330
http://llvm.org/docs/NVPTXUsage.html

**Comment by jfbastien** [comment]
@binji will address in [threads proposal](https://github.com/WebAssembly/threads). Closing.
